### PR TITLE
Secrets Interface & Implementation

### PR DESCRIPTION
**Pull request by jorg3lopez**
# Local and Azure Secrets retrieval

This PR adds the logic to retrieve secrets for our local and Azure environments.

## Testing
- All unit tests should pass

## Issue
#84 - Sending New Born Demographic Data

### CODE REVIEW

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java [hunk 1]
```diff
@@ -85,7 +87,11 @@ protected String requestToken() {
         try {
             senderToken =
                     jwt.generateSenderToken(
-                            CLIENT_NAME, RS_DOMAIN_NAME, "pemKey", CLIENT_NAME, 300);
+                            CLIENT_NAME,
+                            RS_DOMAIN_NAME,
+                            secrets.getKey("AZURE_KEY_NAME"),
```
Comment (jorg3lopez) [inline app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java:92]: Azure key name should go here.

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/javalin/App.java [hunk 2]
```diff
@@ -53,5 +56,10 @@ private static void registerClasses() {
         ApplicationContext.register(OpenApi.class, OpenApi.getInstance());
         ApplicationContext.register(HttpClient.class, ApacheClient.getInstance());
         ApplicationContext.register(AuthEngine.class, JjwtEngine.getInstance());
+        ApplicationContext.register(
+                Secrets.class,
+                ApplicationContext.getEnvironment().equalsIgnoreCase("local")
+                        ? LocalSecrets.getInstance()
+                        : AzureSecrets.getInstance());
```
Comment (jorg3lopez) [inline app/src/main/java/gov/hhs/cdc/trustedintermediary/external/javalin/App.java:63]: I can do an if/else if suggested.

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java (hunk 1 above)
Comment (jorg3lopez) [inline app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java:92]: Should I use a container for this so we check what we got from Azure?

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/localfile/LocalSecrets.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package gov.hhs.cdc.trustedintermediary.external.localfile;
+/**
+ * This Class implements the Secret interface, and it's purpose is to retrieve secrets from our
+ * local environment
+ */
```
Comment (halprin) [inline app/src/main/java/gov/hhs/cdc/trustedintermediary/external/localfile/LocalSecrets.java:5]: Class level JavaDocs should show up right above the `class`, not above the imports.

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java (hunk 1 above)
Comment (halprin) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java:92]: Here's what to put in replacement of `"AZURE_KEY_NAME"`.

```java
var senderPrivateKey = "report-stream-sender-private-key-" + ApplicationContext.getEnvironment();
secrets.getKey(senderPrivateKey)
```

Comment (halprin) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java:92]: You mean test in a Docker container to see if we actually retrieve the secret?  Or...?

File: app/build.gradle [hunk 4]
```diff
@@ -77,6 +77,10 @@ dependencies {
     runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.11.5'
     runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.11.5'
 
+    // azure sdk
+    implementation 'com.azure:azure-security-keyvault-secrets:4.5.4'
+    implementation 'com.azure:azure-identity:1.8.0'
```
Comment (halprin) [inline app/build.gradle:82]: See if this fixes the vulnerability.

```suggestion
    implementation 'com.azure:azure-identity:1.8.1'
```

File: app/build.gradle [hunk 5]
```diff
@@ -77,6 +77,10 @@ dependencies {
     runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.11.5'
     runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.11.5'
 
+    // azure sdk
+    implementation 'com.azure:azure-security-keyvault-secrets:4.5.4'
```
Comment (halprin) [inline app/build.gradle:81]: See if this fixes the vulnerability.

```suggestion
    implementation 'com.azure:azure-security-keyvault-secrets:4.6.0'
```

**Review by halprin — COMMENTED** [review]
This is going to be so sweet!

File: app/build.gradle (hunk 4 above)
Comment (jorg3lopez) [reply app/build.gradle:82]: Trying the update. I meant for the Azure one, not the local ;)

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/localfile/LocalSecrets.java (hunk 3 above)
Comment (jorg3lopez) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/external/localfile/LocalSecrets.java:5]: Good catch!

File: app/build.gradle (hunk 5 above)
Comment (jorg3lopez) [reply app/build.gradle:81]: Good idea, trying the update.

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/azure/AzureSecrets.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * This Class implements the Secret interface, and it is a "humble object" that uses the "Azure
+ * Security Key Vault" and "Azure Identity" dependencies. and it is used to retrieve azure
+ * environment secrets along-side Azure keys from the Azure key vault.
+ */
+public class AzureSecrets implements Secrets {
+
+    private final String keyVaultName =
+            ApplicationContext.getProperty("KEY_VAULT_NAME"); // Verify the env variable name
```
[the remark this answers is not in the document]
Comment (jorg3lopez) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/external/azure/AzureSecrets.java:20]: Got it!

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java (hunk 1 above)
Comment (jorg3lopez) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamLabOrderSender.java]: meant for the Azure env.

**Review by tjohnson7021 — APPROVED** [review]

**Review by halprin — COMMENTED** [review]
Good stuff!  Let's wait a little bit until Azure fixes that vulnerable dependency.  Then we can update to it, and we should be good to go.

**Review by halprin — APPROVED** [review]

**Comment by halprin** [comment]
Waiting on https://github.com/Azure/azure-sdk-for-java/pull/34256 to merge and be released.  And here's the issue: https://github.com/Azure/azure-sdk-for-java/issues/34223.

**Comment by halprin** [comment]
I think we should merge this PR.  I don't think there are vulnerabilities.
1. `io.netty:netty-handler` vulnerability [doesn't apply to Azure's library according to them](https://github.com/Azure/azure-sdk-for-java/issues/18286).
2. `net.minidev:json-smart` is not vulnerable if we specify version `2.4.10` specifically in our `build.gradle` file.  Read below for details on this issue.

I was tired of waiting for Azure to release the fix for https://github.com/Azure/azure-sdk-for-java/issues/34223.  I remember being surprised that adding `implementation 'net.minidev:json-smart:2.4.10'` to the `build.gradle` file didn't work when you all tried it.  Sure enough, Snyk's PR check still complains (duh).  I decided to check two different things though.

When you specify version 2.4.10 and then run `./gradlew app:dependencies`, you'll notice that all of the `net.minidev:json-smart` dependencies are resolving to version _2.4.10_, not the vulnerable 2.4.8!  I even check IntelliJ's list of external library and it shows 2.4.10.  So now I'm starting to think that there is an issue with Snyk's PR check.

Snyk also creates a Gradle plugin to help scan dependencies.  So I spin that up and it _doesn't_ report the vulnerable 2.4.8 dependency!  That tells me that when Snyk runs natively with Gradle through the plugin, it seems to understand that the non-vulnerable 2.4.10 version is being used, but the PR check doesn't figure that out.

So, due to all of this, I think we are safe.  When the new version of `com.azure:azure-identity` comes out, we can update to that and remove the `net.minidev:json-smart:2.4.10` override.

**Pull request merged**